data_preparation/global_AgERA5/download_agera5.py

```diff
@@ -0,0 +1,123 @@
+import os
+import cdsapi
+import concurrent.futures
+
+# Most of the code here comes from Abdelrahman's library.
+# https://github.com/AbdelrahmanAmr3/earthstat
+
+AgERA5_params = {
+    "Maximum_Temperature": ("2m_temperature", "24_hour_maximum"),
+    "Minimum_Temperature": ("2m_temperature", "24_hour_minimum"),
+    "Mean_Temperature": ("2m_temperature", "24_hour_mean"),
+    "Solar_Radiation_Flux": ("solar_radiation_flux", None),
+    "Precipitation_Flux": ("precipitation_flux", None),
+}
+
+
+# @author: Abdelrahman Saleh
+def get_agera5_params(sel_param, start_year, end_year, area=None):
+    if sel_param in AgERA5_params:
+        var, stat = AgERA5_params[sel_param]
+
+        # Construct the return dictionary
+        retrieve_params = {
+            "version": "1_1",
+            "format": "zip",
+            "variable": var,
+            "year": [str(year) for year in range(start_year, end_year + 1)],
+            "month": [
+                "01",
+                "02",
+                "03",
+                "04",
+                "05",
+                "06",
+                "07",
+                "08",
+                "09",
+                "10",
+                "11",
+                "12",
+            ],
+            "day": [
+                "01",
+                "02",
+                "03",
+                "04",
+                "05",
+                "06",
+                "07",
+                "08",
+                "09",
+                "10",
+                "11",
+                "12",
+                "13",
+                "14",
+                "15",
+                "16",
+                "17",
+                "18",
+                "19",
+                "20",
+                "21",
+                "22",
+                "23",
+                "24",
+                "25",
+                "26",
+                "27",
+                "28",
+                "29",
+                "30",
+                "31",
+            ],
+        }
+
+        if stat is not None:
+            retrieve_params["statistic"] = stat
+
+        if area is not None:
+            retrieve_params["area"] = area
+
+        return retrieve_params
+    else:
+        return None
+
+
+def download_agera5_year(cds, year, sel_param):
+    download_path = os.path.join("/path/to/downloads", sel_param)
+    os.makedirs(download_path, exist_ok=True)
+    retrieve_params = get_agera5_params(sel_param, year, year)
+    cds.retrieve(
+        "sis-agrometeorological-indicators",
+        retrieve_params,
+        f"{download_path}/{sel_param}_{year}.zip",
+    )
+
+
+def download_agera5(cds, num_requests, start_year, end_year):
+    # NOTE: $HOME/.cdsapirc needs to contain API key
+    # Example contents of .cdsapirc:
+    # url: https://cds.climate.copernicus.eu/api/v2
+    # key: {cdsapi.uid}:{cdsapi.api_key}
```

Comment:
If we add assertion to be sure that the HOME contain .cdsapric, it will be good.

You can use my CDS API key manger script:
https://github.com/AbdelrahmanAmr3/earthstat/blob/master/earthstat/xES/cds_api_key_manager.py